### ISSUE TITLE
GORM finder performance significantly degraded by Hibernate session size

### ISSUE THREAD

**Issue by timbonicus** [grails/grails-core#10593]
I reported issue #10586 initially for a performance problem in Grails 3.2.8, but focused on the Hibernate FlushMode issue in that ticket. In further testing, there seems to be an issue beyond FlushMode.

The attached project has an integration test (PerformanceSpec) that prints out timing for the same GORM finder:
1. Before anything is loaded in the Hibernate session
2. After loading 50,000 objects
3. After clearing the Hibernate session

The output shows:
```
Finder cost: 2
Loaded 50000 objects into session
Finder cost (after load): 42
Cleared session
Finder cost (empty session): 1
```

I see the same performance hit changing this 3.2.8 project to use `gorm-hibernate4`, but **not** in a Grails 2.5.4 application:
```
Finder cost: 4
Loaded 50000 objects into session
Finder cost (after load): 2
Cleared session
Finder cost (empty session): 1
```

Since the Grails 2.5.4 application and my gorm-hibernate4 test produced different results (while using a nearly identical version of Hibernate), this may imply a GORM issue rather than Hibernate itself.

### Environment Information
- **Operating System**: Windows 10
- **Grails Version:** 3.2.8
- **JDK Version:** 1.8.0_112

### Example Application
[test-hibernate.zip](https://github.com/grails/grails-core/files/930960/test-hibernate.zip)

**Comment by graemerocher** [comment]
This is probably the same as #640

Will verify

**Comment by timbonicus** [comment]
@graemerocher The sample application contains the recommended settings from #640, including the removal of the spring-actuator dependency. Sorry, I should have mentioned that. Thanks for having a look, I hope you can uncover something!

**Comment by graemerocher** [comment]
Can you please provide the Grails 2 sample application you used to produce the same output above.

**Comment by graemerocher** [comment]
Looking at the example if I upgrade to GORM 6.1.1 and use `@ReadOnly` to demarcate the transaction as read-only:

```groovy
@ReadOnly
    def testFinderPerformance() {
```
Then I get:

```
Grails application running at http://localhost:57739 in environment: test
Finder cost: 1
Loaded 50000 objects into session
Finder cost (after load): 1
Cleared session
Finder cost (empty session): 1
```
So it is probably another example of how using properly demarcated transactions is important

**Comment by timbonicus** [comment]
What if the test isn't read-only? This performance issue has made our integration test suite go from 30 minutes runtime in Grails 2.5.4 to over 2 hours in Grails 3.2.8. I'm sure you can understand what a hit that is to developer productivity.

I'm a little alarmed at the WontFix label, and I hope this will receive a little further investigation. Here is the Grails 2.5.4 example.
[test-hibernate-grails2.zip](https://github.com/grails/grails-core/files/935527/test-hibernate-grails2.zip)

**Comment by graemerocher** [comment]
Fix requires upgrading to GORM 6.1.2 when it is released

**Comment by timbonicus** [comment]
Thank you @graemerocher for the big fix!
